inst/www/shared/shiny.js

```diff
@@ -1260,106 +1279,801 @@
         }
 
-        if (clickId)
-          $(img).on('mousedown', createMouseHandler(clickId));
-        if (hoverId) {
-          $(img).on('mousemove', $el.data('hover-func'));
-          $(img).on('mouseout', function(e) {
-            $el.data('hover-func')(null);
-          });
+        // Shift an array of values so that they are within a min and max.
+        // The vals will be shifted so that they maintain the same spacing
+        // internally. If the range in vals is larger than the range of
+        // min and max, the result might not make sense.
+        function shiftToRange(vals, min, max) {
+          if (!(vals instanceof Array))
+            vals = [vals];
+
+          var maxval = Math.max.apply(null, vals);
+          var minval = Math.min.apply(null, vals);
+          var shiftAmount = 0;
+          if (maxval > max) {
+            shiftAmount = max - maxval;
+          } else if (minval < min) {
+            shiftAmount = min - minval;
+          }
+
+          var newvals = [];
+          for (var i=0; i<vals.length; i++) {
+            newvals[i] = vals[i] + shiftAmount;
+          }
+          return newvals;
         }
 
-        if (clickId || hoverId) {
-          $(img).addClass('crosshair');
+        // Set cursor to one of 3 styles. We need to set the cursor on the whole
+        // img instead of the brush dive, because the brush div has
+        // 'pointer-events:none' so that pointer events will pass through to the
+        // img.
+        function setCursorStyle(style) {
+          if (style === 'crosshair') {
+            $el.addClass('crosshair');
+            $el.removeClass('grabbable');
+            $el.removeClass('grabbable');
+
+          } else if (style === 'grabbable') {
+            $el.removeClass('crosshair');
+            $el.addClass('grabbable');
+            $el.removeClass('grabbing');
+
+          } else if (style === 'grabbing') {
+            $el.removeClass('crosshair');
+            $el.removeClass('grabbable');
+            $el.addClass('grabbing');
+          }
         }
+
+        function sendBrushInfo() {
+          var bounds = brush.bounds;
+
+          // We're in a new or reset state
+          if (isNaN(bounds.xmin)) {
+            exports.onInputChange(inputId, null);
+            return;
+          }
+
+          // Transform coordinates of brush to data space
+          var min = offsetToScaledCoords({ x: bounds.xmin, y: bounds.ymin }, opts.brushClip);
+          var max = offsetToScaledCoords({ x: bounds.xmax, y: bounds.ymax }, opts.brushClip);
+
+          // Because the x and y directions of the pixel space may differ from
+          // the x and y directions of the data space, we need to recalculate
+          // the min and max.
+          var coords = {
+            xmin: Math.min(min.x, max.x),
+            xmax: Math.max(min.x, max.x),
+            ymin: Math.min(min.y, max.y),
+            ymax: Math.max(min.y, max.y),
+            '.nonce': Math.random()
```

Comment:
Maybe brush shouldn't get a nonce? This doesn't feel "eventy" to me, whereas click/dblclick and arguably hover do.